### ISSUE TITLE
Proposal for nonintrusive pipeline parallelism

### ISSUE THREAD

**Issue by sogartar** [nod-ai/shark-ai#2045]
We would like the model definition to not be burdened with pipeline parallelism (PP) logic. Making PP orthogonal to the model should yield faster model development.
There is a plausible path to achieve this by utilizing PyTorch's traced graph representation. The graph would be expected and pipeline splits inserted to break the model into multiple stages. It seems this is what is done in the [torch.distributed.pipelining](https://docs.pytorch.org/docs/stable/distributed.pipelining.html) submodule.

E.g.
```python
import torch
from torch.nn import Linear
from torch.distributed.pipelining import pipeline, SplitPoint
import iree.turbine.aot


class Model(torch.nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.emb = torch.nn.Embedding(10, 3)
        self.layers = torch.nn.ModuleList(Linear(3, 3) for _ in range(2))
        self.lm = Linear(3, 3)

    def forward(self, x: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
        x = self.emb(x)
        for layer in self.layers:
            x = layer(x)
            x = x + y
            y = x * x
        z = self.lm(x)
        return x, z


model = Model()


# An example input.
x = torch.LongTensor([1, 2, 4, 5])
y = torch.rand([4, 3])
model(x, y)

# Split in a pipeline of 2 stages where the split happens at the start of layer.1.
pipe = pipeline(
    module=model,
    mb_args=(x, y),
    split_spec={
        "layers.1": SplitPoint.BEGINNING,
    },
)

print("Original model:")
print(model)
print("PP model:")
print(pipe)
```

```
Original model:
Model(
  (emb): Embedding(10, 3)
  (layers): ModuleList(
    (0-1): 2 x Linear(in_features=3, out_features=3, bias=True)
  )
  (lm): Linear(in_features=3, out_features=3, bias=True)
)
```
```
PP model:
GraphModule(
  (submod_0): GraphModule(
    (emb): InterpreterModule()
    (layers): Module(
      (0): InterpreterModule()
    )
  )
  (submod_1): GraphModule(
    (layers): Module(
      (1): InterpreterModule()
    )
    (lm): InterpreterModule()
  )
)



def forward(self, x, y):
    submod_0 = self.submod_0(x, y);  x = y = None
    getitem = submod_0[0]
    getitem_1 = submod_0[1];  submod_0 = None
    submod_1 = self.submod_1(getitem, getitem_1);  getitem = getitem_1 = None
    getitem_2 = submod_1[0]
    getitem_3 = submod_1[1];  submod_1 = None
    return (getitem_2, getitem_3)
```

The result is a torch Module of type [torch.distributed.pipeline.Pipe](https://github.com/pytorch/pytorch/blob/284b7190054686e68d9cc683b6ce43e45dd22338/torch/distributed/pipelining/_IR.py#L524), which holds a [graph module](https://github.com/pytorch/pytorch/blob/284b7190054686e68d9cc683b6ce43e45dd22338/torch/fx/graph_module.py#L417) that orchestrates/wires the calling into the pipeline stage submodules. This graph is used to generate the `forward` function in the example.
We can export each submodule as a separate stage into a different MLIR function. We want a function per stage if at runtime we are going to control the pipeline concurrency from outside of IREE functions. Most likely this is what we want. For more details see https://github.com/nod-ai/shark-ai/issues/1562. We would need to programmatically inspect the generated graph and stage submodules.
We could potentially not use the already existing functionality in torch for module splitting and implement it ourselves as it is not clear how stable it is going to be in the future. This would require patching the module to mark split points, tracing and then transforming the graph into stages as separate submodules. Even if we use the `torch.distributed.pipeline` we still would need to patch our modules to insert tensor transfers between devices when transitioning between stages. It is good that `torch.distributed.pipeline.pipeline` does not insert any new `torch.device` movements. This is done separately.
To make exporting to MLIR not be burdensome we would need to trace through the stages one by one to get at meta-tensor info for the interstage arguments/results. Then we inspect how arguments/results are propagated/wired between stages in order to seed the export for the next stage. It seems like that we can propagate also dynamic dimension info.
E.g.
```Python
dyn_dim = torch.export.Dim("x0")
dynamic_shapes = {
    "x": {0: dyn_dim},
    "y": {0: dyn_dim}
}
submod_0_export = torch.export.export(pipe.split_gm.submod_0, args=(x, y), dynamic_shapes=dynamic_shapes)
```

This is one of the results of the submodule:
<img width="527" height="133" alt="Image" src="https://github.com/user-attachments/assets/5446027d-fcf7-45a8-a577-8cbb2c1a3d03" />

We want to also export this information in some format alongside our models, so that users (shortfin) can inspect how to wire the pipeline as they will be the ones orchestrating the concurrency. Potentially for convenience we can have additional MLIR functions exported that prepare pipeline stage arguments from the initial arguments and the results of all previous stages. They would also maintain a live set of tensors that would be required for later stages, making sure that interstage tensors don't live longer than they should.

There are some complications with our custom tensor types like sharded tensors. When they are a part of an export signature they will be flattened into the constituent `torch.Tensor`s and will have to be reconstructed when needed using the [registered functions](https://github.com/nod-ai/shark-ai/blob/17ee1991ce53d5fe6cfc55f910f17edf68f80ada/sharktank/sharktank/types/tensors.py#L1820-L1825) with PyTorch.

Our paged KV cache right now is passed to all LLM block/layers as a single tensor and it stays as such during all [computations](https://github.com/nod-ai/shark-ai/blob/17ee1991ce53d5fe6cfc55f910f17edf68f80ada/sharktank/sharktank/layers/paged_attention.py#L276). In the PP case it will be a sharded tensor, split along the block/layer dimension. It will still be passed to each stage as such and then all appropriate ops would need to be implemented to operate on the split tensor. We would have to somehow specify that the KV cache does not need insertion of device transfer ops for the cache.
E.g.
```python
pipe = pipeline(
    module=model,
    mb_args=(x, y),
    args_spec={ "x": NoInterStageAutoTransfer }, # <-
    split_spec={
        "layers.1": SplitPoint.BEGINNING,
    },
)
```

**Comment by Alex-Vasile** [comment]
My biggest issue is that [torch.distributed.pipelining](https://docs.pytorch.org/docs/stable/distributed.pipelining.html) is in an alpha. That doesn't seem ready to become load bearing for us.

**Comment by sogartar** [comment]
> My biggest issue is that [torch.distributed.pipelining](https://docs.pytorch.org/docs/stable/distributed.pipelining.html) is in an alpha. That doesn't seem ready to become load bearing for us.

The we roll out our own pipeline splitting transformation.
